# Лабораторная работа №1

Цель работы - изучить постановку антагонистической игры двух лиц в нормальной форме; найти решение игры за обоих игроков в смешанных стратегиях (стратегическую седловую точку).

## Постановка задачи и методические указания

Для игры, заданной матрицей $c_{ij}$, требуется найти оптимальные смешанные стратегии обоих игроков, сведя матричную игру к задаче ЛП (прямой для одного игрока и двойственной для другого).

Задачи ЛП следует решать симплекс-методом, приводя начальные, промежуточные и конечные симплекс-таблицы. По окончании алгоритма полученные решения необходимо проверить на допустимость.

## Ход работы


Зададим матрицу стратегий:

In [1]:
from IPython.core.display import display, HTML, Latex

C = matrix(SR, 4, 5, [8, 1, 17, 8, 1, 12, 6, 11, 10, 16, 4, 19, 11, 15, 2, 17, 19, 6, 17, 16])
C

[ 8  1 17  8  1]
[12  6 11 10 16]
[ 4 19 11 15  2]
[17 19  6 17 16]

Сформулируем задачу линейного программирования для решения симплекс-методом для игрока $A$:

In [2]:
A = C.transpose()
b = [1] * A.nrows()
c = [1] * A.ncols()

P = InteractiveLPProblem(A, b, c, constraint_type=[">="] * len(b), problem_type="min", variable_type='>=')

Решим задачу ЛП симплекс-методом:

In [3]:
P = P.standard_form()
Latex(P.run_simplex_method())

<IPython.core.display.Latex object>

In [4]:
D = P.final_dictionary()

W = -D.objective_value()
g = 1/W

solution = D.basic_solution()

x = [i * g for i in solution]
print(x)

[553/3684, 1555/3684, 773/3684, 803/3684]


Таким образом, оптимальная смешанная стратегия игрока $A$: $\left(\frac{553}{3684}, \frac{1555}{3684}, \frac{773}{3684}, \frac{803}{3684}\right)$.

Сформулируем задачу ЛП для игрока $B$:

In [5]:
B = C
b = [1] * B.nrows()
c = [1] * B.ncols()

P = InteractiveLPProblem(B, b, c, constraint_type=["<="] * len(b), problem_type="max", variable_type='>=')

In [6]:
P = P.standard_form()
Latex(P.run_simplex_method())

<IPython.core.display.Latex object>

In [7]:
D = P.final_dictionary()

Z = D.objective_value()
h = 1/Z

solution = D.basic_solution()

v = [i * h for i in solution]
print(v)

[233/3684, 185/921, 719/1228, 0, 277/1842]


Таким образом, оптимальная смешанная стратегия игрока $B$: $\left(\frac{233}{3684}, \frac{185}{921}, \frac{719}{1228}, 0, \frac{277}{1842}\right)$.